### ISSUE TITLE
Android 64 bit support

### ISSUE THREAD

**Issue by btelman96** [mixpanel/mixpanel-unity#56]
I was going though libraries I use in a Android unity project and noticed that Mixpanel does not appear to have Armv8 (64 bit) support, which will be required for all Google Play store apps this August to at least publish a new version.

File: libMixpanelSDK.so (currently only shows in the jni armv7 and x86 folders when inspecting my apk)

Unity 2017.4.17f1
IL2CPP 64 Bit build
Mixpanel Version 1.3.1

**Comment by chinhanglun** [comment]
I would also like a 64 bit build too.

Any advice on how to build it myself will be appreciated.

**Comment by chinhanglun** [comment]
macOS 10.13.6
Cmake 3.9.1
android-ndk-r10e

I changed line 563 of native/build/build_all.py to `android_abis = ('arm64-v8a')`

Then ran:

`python build_all.py --no-test --platforms android`

I am getting errors like:
```

../../libmixpanel.a(platform_helpers.cpp.o): In function `mixpanel::detail::PlatformHelpers::collect_automatic_properties()':
platform_helpers.cpp:(.text._ZN8mixpanel6detail15PlatformHelpers28collect_automatic_propertiesEv+0x6c): undefined reference to `__system_property_get'

```
...

```
Traceback (most recent call last):
  File "build_all.py", line 599, in <module>
    main()
  File "build_all.py", line 583, in main
    build.build()
  File "build_all.py", line 311, in build
    subprocess.check_call(['cmake', '--build', '.', '--config', self.configuration, '--', '-j8'])
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--', '-j8']' returned non-zero exit status 2
```

Why is this happening?

**Comment by chinhanglun** [comment]
Turns out that the method __system_property_get were removed from android NDK r10e but it was put back into NDK r11c and so building with the latter works.

Link to the .so file below for anyone can't be bothered to build it.

**Comment by rocktavious** [comment]
@btelman96 and @chinhanglun I've done a Mixpanel rewrite that i'd be interested in having you try out to see if you can use it with the version of unity + android you need.  https://github.com/rocktavious/mixpanel-unity

If it does work for you then it could go along way to validating my fork to be merged back upstream.

Here is the PR I've started to get mixpanel to fix their library https://github.com/mixpanel/mixpanel-unity/pull/57

**Comment by vzlomvl** [comment]
Hello @zihejia . Does MixPanel support Unity Android x64 build? If no, when does MixPanel plan to add support of IL2CPP 64 Android build? From August 2019 Google Play will not allow uploading builds without x64 support and we will be forced to stop using MixPanel to upload a new build if support doesn't appear.
Thanks.

**Comment by patedit** [comment]
hi @vzlomvl ! we are planning on cutting a new release by August that adds a bunch of new functionality, including android 64 bit support!

**Comment by memidgety** [comment]
I was wondering on the status of the 64 bit release as tomorrow is August 1st, when android will require a 64bit version of apps. We will not be able to push Mixpanel unity apps to the app stores during this time without a 64bit DLL

**Comment by rocktavious** [comment]
@memidgety You can test out the new functionality by using this branch https://github.com/mixpanel/mixpanel-unity/tree/release/2.0 which should fix your problems with compiling unity to 64 bit.

**Comment by manituan** [comment]
We've been trying to test out the 2.0 version on our Unity project but we can't see the events showing up in our mixpanel dashboard. We've set up the tokens in the SO and we call the Mixpanel.Track method as normal. No errors in the console and we erased the PlayerPrefs. Is there anything I'm missing? Thanks

**Comment by manituan** [comment]
After debugging it for a while I found that DoFlush doesn't do anything because Mixpanel.Batches is empty, but the Buffer wasn't so I forced PrepareBatches at the beginning of DoFlush filling the dictionary and now I get to MixpanelManager.BuildRequest. It says it's successful but I still I can't get my events to show on the mixpanel dashboard. 

Hope this is helpful. Thanks.

**Comment by rocktavious** [comment]
@manituan i'm on vacation untill Aug 12 but i can look into this when i get back.  The track calls should just enqueued into the buffer and then every 60 seconds (due to the default flush interval) they should get batched up into groups of 40 and sent to the Mixpanel API.   If you turn on the "debug" flag do you get any output of errors or such.  I did recently make some changes around this I wonder if you go back to the version of 2.0 from here https://github.com/mixpanel/mixpanel-unity/commit/d9b9c5af31cb4d964edf286c47eb0bcd6bb287ad if that fixes you - then that means i made the problem in commit https://github.com/mixpanel/mixpanel-unity/commit/164f79a25a7c9d5acc85c00dfa88b34c6d4c1ae9 likely

**Comment by manituan** [comment]
@rocktavious Thanks for you answer. 
By reverting to that commit BuildRequest is called without me having to edit the code. That's good. But the events still don't show up in our mixpanel dashboard. 
I have always mixpanel debug enabled and I can see all messages composed and then sent. There are no errors. Our tokens are correct, I've double checked. 

This is blocking our Android release :( 
Cheers!

**Comment by kotredes** [comment]
I too faced this problem, and I think I have a fix/workaround. (Side note: it's really pretty disappointing that even with 6 months of heads up, there wasn't an actual released update of the mixpanel unity plugin that you can use to submit Android builds to Google play)

Anyways, based on my reading of the code:  When you track an event, it gets put into a “Buffer” object.  Later, some other function (PrepareBatches) is supposed to move those Buffered objects to the “Batches” object.  The “every 60 second flush to mixpanels server” call sends what’s in “Batches” to their server.    Thing is, that “PrepareBatches()” function is only called by the “SaveBatches()” function, which is only called in three cases: The app is quit, The app is backgrounded, or when you manually call Mixpanel.Flush().   Note: The every-60-second-auto-flush logic does *not* use Mixpanel.Flush(), and thus does not automatically call SaveBatches().  

The workaround I used was adding a “Mixpanel.SaveBatches()” call as the very first line in MixpanelManager’s DoFlush() function.    Without this, I observed that messages only got sent to mixpanel when reopening the app after backgrounding/quitting it.  With this change, messages get sent every minute, as expected.

**Comment by rocktavious** [comment]
@kotredes Thanks for looking into this.  I'm almost back from my vacation so i can dig in more but this really helps alot and likely was just something i missed in my recent rewrite of the batching/queueing.  As to your side note - I don't work for Mixpanel (and only after my rewrite of this library - have i started to do contract work for them to help improve this library)  I feel you pain but I'm doing my best to help out the community using this mixpanel-unity library.  This library started out as an opensource contribution and while it is under the mixpanel github organization now mixpanel does not have any unity developers on staff (this is why they are contracting out or looking for help from the unity community to help keep this integration afloat)  I hope that my personal time spent on rewriting the library helps you along as it has for me.  Thanks again!

**Comment by manituan** [comment]
@kotredes I did the same thing, but my events dont show up in the dashboard. Log is enabled and I can see the events being sent. IsTracking is true. The mixpanel setting asset is filled with the correct release and debug tokens, but when I update the live view I see nothing. 
Can you see them arriving to the mixpanel dashboard on your end?
If I revert to v1.3 it works as normal.

**Comment by chinhanglun** [comment]
If all you want is to be able to submit to Google Play then just get the .so file I linked above which is the plugin built for arm64 and put it in Plugins/Android/libs/arm64-v8a in your Unity project.

This is what I have done and it work or you can just build the plugin yourself for your desired architecture.

**Comment by rocktavious** [comment]
@manituan @kotredes I made this change https://github.com/mixpanel/mixpanel-unity/commit/8ef8559448110d14a67dbc2ccf075f4a7bf8a02e which should result in the intended code flow I meant when writing this.  I setup a project in an editor and submitted 6 track calls then waited 60 seconds and I see the events show up in the mixpanel dashboard.  I also built a standalone exe version of it and submitted 4 track class and then waited 60 seconds and I see the events show up in the mixpanel dashboard.
![Proof](https://user-images.githubusercontent.com/735015/62776520-6ed3f980-ba70-11e9-9e65-c240df357272.PNG)

Let me know what you think.  @manituan if you are still having problems using the commit linked above from the 2.0 branch can you please validate the tokens you have in unity vs the token in your mixpanel project.  Make sure your using the "Token" and not the "API Key" or "API Secret" for the live view of the project you are not seeing them show up in - Thanks!

**Comment by manituan** [comment]
@chinhanglun That works. Thank you.
 
@rocktavious Thanks for trying to make this work while you're on holidays. I appreciate it. 
I'm afraid latest doesn't work for me either. By that I mean that I still can't see the events showing up in the mixpanel live dashboard. 
I also thought I could have written the tokens wrong, so I double checked it. But I'm 100% sure now I'm using the same tokens as in the 1.3 version. 

Cheers!

**Comment by rocktavious** [comment]
@manituan Did you use https://github.com/mixpanel/mixpanel-unity/commit/8ef8559448110d14a67dbc2ccf075f4a7bf8a02e ?  That change is nearly exactly what @kotredes described - just slightly different as I wanted it to be using the same API mechanisms that a user would use to "flush" the queue such that if you used the API call or the 60 second timer used the API one would expect the same behaviour.

I just want to be 100% sure that the latest 2.0 branch doesn't work for you on Android before I request access to an Android device from Mixpanel to dig deeper into this (I've never done Android development before so it might take me some time to really dig in but i'd want to make sure the new 2.0 branch works for everyone so we can cut over to it as its a much much more simplified version of the library)

**Comment by manituan** [comment]
@rocktavious Yes, I got latest. That change is exactly what I did to fix it one week ago. I can see in the log the events being thrown but they don't get to the mixpanel dashboard.

To test this changes I'm working on a standalone environment. So it's not Android related. The reason why I need this working is because Android, but now it doesn't work for me regardless the platform I'm using. Hope it makes sense.

It would be useful to know if someone else can get their events in the mixpanel dashboard on latest.

**Comment by rocktavious** [comment]
@manituan Hmm interesting - because I just tested it after i made the change in a standalone build and in the image above you can see the events in the live view.
What version of Unity are you using?  Also can you test a clean project?  Like start a fresh Unity 2018.4 project and unpack the mixpanel demos and see if that works for you?  Thanks!

**Comment by kotredes** [comment]
For what it's worth, we *do* see our events on the mixpanel dashboard.

That being said, I've noticed a few changes from the way the older SDK worked that might be causing you to not see your events.

Most specifically, it looks like my new events are showing with an Operating System of "Other" (Previously, it was "Android"). 

@manituan, If your dashboard was filtering on Operating System, maybe that's why you're not seeing the events?  (Assuming you are seeing the "[Mixpanel] Sending Request" log when you turn on "Show Debug", then my suggestion, if you haven't already done so, is to check out the "Live View" section of mixpanel.com, and see if you can see the events there.)

Another minor thing I happened to notice, but might trip someone up: 2.0 doesn't auto-populate the $unity_lib_version People property the way the old SDK did.   

(Note: I haven't attempted to audit the events for any/all differences between the SDK version, so there may be others subtle things that have changed, I just happened to see these two particular changes while poking at things)

Hope this helps!

**Comment by rocktavious** [comment]
@kotredes This is great information - one of the problems with the rewrite is that not everything property wise was easily gettable from pure c# so its been a bit of a task to make sure everything comes out as expected as the library is no longer using native c++ code - there were also some sacrifices that needed to be made as certain things currently will not come out exactly the same from pure c#.  I tried to do my own Audit from a standalone perspective as I do not have an Android device to test with.

It would be great help if you could audit the changes in the properties for events between the old and the new lib as it might provide some insight into what I would need to change in regards to Android.

I will try to dig into this more to figure out why by cross referencing the old version of the lib and the new version but an Audit will make it much more plain to see.  Thanks for helping test out this rewrite - I have hope this rewrite will make it easier for Mixpanel to maintain this library in the future.

**Comment by rocktavious** [comment]
@kotredes It looks like based on my auto properties information from Mixpanel `$unity_lib_version` was changed to just `$lib_version` which makes it in-sync with the rest of the mixpanel SDK's.

I modified the 2.0 version of library and submitted a track call using 
`$unity_lib_version` shows up as a "user property`instead of a Mixpanel Property
![MixpanelPropertyTest](https://user-images.githubusercontent.com/735015/62794799-a22b7e00-ba9a-11e9-853c-4eb9a1cc1dfa.PNG)

But you are correct we are not sending through `$lib_version` as an auto property on people calls unless its iOS or Android which is likely a mistake - I'll find out.

As for the "Operating System" auto property i'm using https://docs.unity3d.com/ScriptReference/SystemInfo-operatingSystemFamily.html which states `This API is used by UI code to distinguish between desktop OS families with different UI conventions. For all non-desktop platforms it will currently return OperatingSystemFamily.Other.`  So I'll have to make a change for iOS and Android  - I will look into fixing this next week.

If you could still get me an audit of the differences between Android in 1.3.1 and 2.0 I would be very thankful!

**Comment by kotredes** [comment]
Re: lib version:  Sorry, I guess I left off some emphasis, that might've caused some confusion.

$lib_version is what I see used for *Tracking* events, but $unity_lib_version is what I see being used for *People* properties . (Also called "Engage", I think?)

Take a look at this pre-2.0 code: https://github.com/mixpanel/mixpanel-unity/blob/e2e3b80f7fcbd7f1c397afbc1ceeccf775cf85ed/native/source/mixpanel/detail/mixpanel.cpp#L536


As for OperatingSystem issue. I don't see anything obvious wrong with the 2.0 code...so I'm kinda confused why I was seeing "Other", but I was!

<img width="796" alt="Screen Shot 2019-08-09 at 9 53 32 AM" src="https://user-images.githubusercontent.com/285710/62795442-a13f2000-ba8b-11e9-8d32-bee70dc1496d.png">

I'm guessing this is some Unity-specific weirdness with that call.. (maybe because the android device I'm using is ancient?) . Anyways,  Looks like the old mixpanel SDK cheated and hard-coded "Android"?

**Comment by rocktavious** [comment]
@kotredes Ahh i wonder if that `$lib_version` or `$unity_lib_version` was just missed in some internal change.  I'll do some more testing locally but i think it supposed to be `$lib_version` everywhere (which is also not the case in 2.0).  I will also double check with my Mixpanel contact.

As for the OperatingSystem issue - it was a mis read on the unity docs from me and I know exactly how to fix it.

If you see anything else out of order property wise please please let me know I want to get it all fixed.

**Comment by kotredes** [comment]
One other thing I noticed, (which might explain my not seeing the unity_lib_version get updated)

There's a couple :  
  #if UNITY_ANDRIOD
That are probably intended to be 
  #if UNITY_ANDROID

https://github.com/mixpanel/mixpanel-unity/blob/release/2.0/Mixpanel/Mixpanel.cs#L78
https://github.com/mixpanel/mixpanel-unity/blob/release/2.0/Mixpanel/Mixpanel.cs#L107

**Comment by rocktavious** [comment]
@kotredes ok i fixed those if defs and auto properties in https://github.com/mixpanel/mixpanel-unity/commit/da8a8e603092e49bb29ffa13226417808db5e169 on the 2.0 branch

**Comment by rocktavious** [comment]
@chinhanglun @kotredes or @manituan Has the latest 2.0 branch been working for you guys now after the latest round of fixes?

**Comment by manituan** [comment]
@rocktavious I got latest this morning and I had to revert last commit due compiler errors in MixpanelSettingsEditor.cs (I removed the ";").
After that I can see my events in the mixpanel dashboard. 
There is one issue though, but I don't know if it's due 2.0. The events show up as if they were sent 1 hour ago.
Cheers!

**Comment by manituan** [comment]
@rocktavious I'm now working on a different Windows PC from last week (it used to be a laptop). All my events get through now, but when I run the project from my laptop they don't. That explains why I couldn't see my events in the live dashboard last week. I don't know yet the reason behind it.

**Comment by rocktavious** [comment]
2.0 has been released!
